Support for backwards compatible rector rules and version scoping of rules

bbrala opened the pull request:
## Description
Adds AbstractDrupalCoreRector base class to enable backward compatibility. It also skips Rector rules if the current Drupal version is less than the introduced change.

Based on the work  by @mglaman on #222 and refactored since we now have `DeprecationHelper` in core.

This change introduced a number of new things:

### Configuration interface for BC enabled rules
`DrupalRector\Contract\VersionedConfigurationInterface`:  If an configuration implements this interface the developer can supply a version for the configured rule. With this drupal-rector can identify the following things:
1. What version does this rule apply to
2. What version needs a BC fix
3. What version just needs the normal fix.

### Concrete configuration for rules with no own configuration
`DrupalRector\Rector\ValueObject\DrupalIntroducedVersionConfiguration`: A configuration valueobject that only contains a version. If a rector does not need any configuration itself, the developer can use this valueobject to supply version information about the rector.

### Base class for BC rule rectors
`AbstractDrupalCoreRector`: This class can be used as a base for BC fixes. It will check the Drupal version, the configuration object for applicable version and either skip the rule, run in in BC mode, or just run it normally. 

#### Supported node types
Currently the only type of node it applies to is `CallLike`. When new rules arrise that might need to apply to other type of nodes. But lets add those as they come up.

```php
// Create a backwards compatible call if the node is a call-like expression.
if ($node instanceof Node\Expr\CallLike && $result instanceof Node\Expr\CallLike) {
  return $this->createBcCallOnCallLike($node, $result, $configuration->getIntroducedVersion());
}
```

CallLike is a one of the following type of calls:
```php
PhpParser\Node\Expr\FuncCall::class
PhpParser\Node\Expr\MethodCall::class
PhpParser\Node\Expr\New_::class
PhpParser\Node\Expr\NullsafeMethodCall::class
PhpParser\Node\Expr\StaticCall::class
```

#### Future optimization with SkipVoters
In the future we might want to add a SkipVoterInterface so we can determine that a rule does not apply earlier in the process. BUT this means if old code is introduced into a project at a later time, that will not be removed. So not sure we even want that.

#### Rector to remove BC code paths
`DrupalRector\Rector\Deprecation\RemoveDeprecationHelperRector`: This rector will remove all BC calls using `DeprecationHelper` from the code if the Drupal version used is higher than the version check in the code. For now it is only ran in a UnitTest, but as we get closed to Drupal 11, we might want to enable this.


## To Test
- Unit tests
- Functional tests

## Drupal.org issue
https://www.drupal.org/project/rector/issues/3350886

bbrala committed:
Initial work based on mglaman's MR and the release of DeprecationHelper

bbrala committed:
Fix phpstan error and add Drupal::VERSION phpstan error to baseline

bbrala committed:
Update d9 functional test for BC rule as test

bbrala committed:
Add use statement

mglaman commented on phpstan-baseline.neon:
```diff
@@ -35,6 +35,11 @@ parameters:
+			message: "#^Else branch is unreachable because ternary operator condition is always true\\.$#"
+			count: 1
+			path: src/Rector/AbstractDrupalCoreRector.php
+
```
If we add the following to `phpstan.neon` this won't be reported:

```
	dynamicConstantNames:
		- Drupal::VERSION
```

See https://github.com/mglaman/phpstan-drupal/blob/main/extension.neon#L14-L15

bbrala replied:
Fixed

mglaman commented on src/Rector/AbstractDrupalCoreRector.php:
```diff
@@ -0,0 +1,52 @@
+{
+    public function refactor(Node $node)
+    {
+        $drupalVersion = substr(\Drupal::VERSION, -4) === '-dev' ? str_replace(['.x-dev', '-dev'], '.0', \Drupal::VERSION) : \Drupal::VERSION;
```
There is no penalty to run `str_replace` on a non-matching string, so we can simplify this logic and make it easier to read

bbrala replied:
Fixed

mglaman reviewed (commented):
Some initial thoughts, brain couldn't upload easily right now.

bbrala committed:
Add \Drupal::VERION to phpstan.neon to silence error.

 	dynamicConstantNames:
    		- Drupal::VERSION

bbrala committed:
Simplify $drupalVersion normalization.

mglaman commented on src/Rector/AbstractDrupalCoreRector.php:
```diff
@@ -0,0 +1,52 @@
+    public function refactor(Node $node)
+    {
+        $drupalVersion = str_replace(['.x-dev', '-dev'], '.0', \Drupal::VERSION);
+        if (version_compare($drupalVersion, $this->getVersion(), '<')) {
```
My brain says: shouldn't `.0` be `.9999` but I don't know. I guess its best to assume `.0` because we don't know where or what the `dev` is at. It could be an old HEAD. For GitLab CI/DrupalCI `drupal/core` isn't always a tagged release, will this break project update bot because it'll skip a lot of things?

bbrala replied:
Core policy defines that depreactions can only be added in minors, see https://www.drupal.org/about/core/policies/core-change-policies/allowed-changes#minor

So lets see:
Depraction introduced in 10.1.0 should evalutate to false on anything on or after that version. This means that lower than 10.1.0 is correct. Since if you are running on 10.1.x you will evaluate to false.

mglaman replied:
Ah, yeah, you're right :)

mglaman commented on src/Rector/Deprecation/RemoveDeprecationHelperRector.php:
```diff
@@ -0,0 +1,97 @@
+$shorter_pass = \Drupal::service('password_generator')->generate(8);
+CODE_AFTER
+            )
+        ]);
```
I don't think this matches. It's a Rector to Rector our replacements of `backwardsCompatibleCall`?

mglaman replied:
Or is this a Rector to revert BC calls when they're no longer needed?

bbrala replied:
Yes this rector is to remove BC calls when you no longer need those. What if we are at Drupal 11 -> 12 cycle, then we need to probably remove Drupal 10 -> 11 BC calls at some point.

Also if a new major version of your module is released this could help you remove all the BC calls easily if you want to drop a major version

bbrala replied:
But you are also right. This example is not correct and need to be updated with a depracatedhelper call to a non-deprecated helper call. Oops.

bbrala replied:
I wil fix this while rebasing and using a true Drupal 10 deprecataion

bbrala replied:
Resolved

bbrala committed:
Start work on refactoring to use d10

This also means we need to adjust to a configurable interface, we could choose to ignore that for now, and do that in the next step

bbrala committed:
Make interface more genera

bbrala committed:
Refactor to always use Configurable rule for version targeting

bbrala committed:
Fix docblock

bbrala committed:
Add DrupalIntroducedVersionConfiguration for rules that only need to be targetted by version but have no configuration

bbrala committed:
fix expected value of fixture

bbrala committed:
Merge pull request #255 from palantirnet/feature/bc-rules-with-configuration

Refactored BC rules with configuration

bbrala committed:
Fix functional test for user_password

bbrala committed:
Fix RemoveDeprecationHelperRector inline example

bbrala committed:
Fix user_password example by removing namespace

bbrala committed:
Delete user_password from d8 example since that is not a versioned rector

bbrala committed:
Update RemoveDeprecationHelperRectorTest.php

bbrala committed:
Rename to DeprecationHelperRemoveRector and support full versions instead of only majors

bbrala committed:
Better example for DeprecationHelperRemoveRector

bbrala committed:
More helpfull comments and readme around Drupal 10 support

mglaman commented on README.md:
```diff
@@ -12,6 +12,11 @@ Automate fixing deprecated Drupal code.
+* The 0.18.0 and higher releases of drupal-rector will include Rector 0.18+. The upgrade path should be as simple as re-copying the configuration file. `cp vendor/palantirnet/drupal-rector/rector.php`.
+
+> [!WARNING]
+> Drupal rector does not support Drupal 10.0 specifically because this will break support for backwards compatible fixes for which the helpers were introduced in Drupal 10.1. Older and newer versions of Drupal are supported.
```
This confused me, what specifically was changed here? Why can't we just fix them as non-backward compatible like previous version?

bbrala replied:
Yeah we can, but question is if we want that split. Was also thinking about update not. But I'll just run 10.1 and adjust changes to the info.yml accordingly.

 So I will remove this part

mglaman commented on src/Rector/AbstractDrupalCoreRector.php:
```diff
@@ -0,0 +1,90 @@
+                continue;
+            }
+
+            // Check if Drupal version and the introduced version support backward
+            // compatible calls. Although it was introduced in Drupal 10.1 we
+            // also supply these patches for changes introduced in Drupal 10.0.
+            // The reason for this is that will start supplying patches for
+            // Drupal 10 when 10.0 is already out of support. This means that
+            // we will not support running drupal-rector on Drupal 10.0.x.
+            if (version_compare($drupalVersion, '10.1.0', '<') || version_compare($configuration->getIntroducedVersion(), '10.0.0', '<')) {
+                return $result;
+            }
```
Ah I think I get it. We're going to replace 10.0.x code with backwards compatibility introduced in 10.1.x because this will be run using versions of Drupal greater than 10.0.x (in bulk)

Also, ugh. I realize this means modules must require a minimum of 10.1.x if they use it. They cannot do a `^9.5 || ^10 || ^11` bridge. Oh well

bbrala replied:
Yeah we do unfortunately. So it would be:

^9.5 | ^10.1 | ^11

But that's not for local code perse, I'll fix updatebot to do that though

mglaman commented on src/Rector/Deprecation/FunctionToStaticRector.php:
```diff
@@ -17,12 +18,12 @@
+class FunctionToStaticRector extends AbstractDrupalCoreRector
 {
     /**
      * @var array|FunctionToStaticConfiguration[]
      */
-    private array $configuration;
+    protected array $configuration;
```
Does this still need to be defined if the parent has it? Ah, for typing analysis it does.

bbrala replied:
Yeah it a more specific type indeed.

mglaman commented on tests/src/Rector/Deprecation/DeprecationHelperRemoveRector/fixture/user_password_keep.php.inc:
This is the same as `user_password.php.inc`. It says `_keep` so I assumed it should keep the backwards compatible call. But it doesn't

bbrala replied:
oops

mglaman commented on tests/src/Rector/Deprecation/DeprecationHelperRemoveRector/fixture/user_password_keep_simplefqcn.php.inc:
What oes this test do? it says `simplefqcn` but I don't  see it doing anything about fully qualified class names.

bbrala replied:
hmm, that was an issue at some point. Forgot to clean up

bbrala committed:
Fix feedback

bbrala committed:
rename file user_password_keep.php.inc

mglaman commented on tests/src/Rector/Deprecation/DeprecationHelperRemoveRector/fixture/user_password_keep.php.inc:
```diff
@@ -0,0 +1,27 @@
+
+function simple_example()
+{
+    $password = DeprecationHelper::backwardsCompatibleCall(\Drupal::VERSION, '10.5.0', fn() => user_password(), fn() => \Drupal::service('password_generator')->generate());
```
👍 so if 10.5.0 is released or when 11.0.0 is released, this test _should_ fail.

bbrala replied:
Hehe technically, but unit tests don't use a real install but a shim for the version. ;)

mglaman reviewed (approved):
I'm going to approve! This looks great. It's a huge improvement to Rector. Before it just fixed anything regardless the version of Drupal.

I guess most rules _still_ do, but now we're able to properly scope rules based on the version of Drupal core! Folks can use drupal-rector without being on the latest minor version of Drupal core!
